Wikipedia/Code/WMFURLSchemeHandler.m

```diff
@@ -0,0 +1,494 @@
+#import "WMFURLSchemeHandler.h"
+#import <WMF/NSURL+WMFSchemeHandler.h>
+#import "Wikipedia-Swift.h"
+#import <WMF/WMFImageTag.h>
+#import <WMF/WMFImageTag+TargetImageWidthURL.h>
+#import <WMF/NSString+WMFHTMLParsing.h>
+#import <WMF/WMFFIFOCache.h>
+#import "MWKArticle.h"
+
+NSString *const WMFURLSchemeHandlerScheme = @"wmfapp";
+NSString *const WMFSchemeHandlerArticleSectionDataBasePath = @"articleSectionData";
+NSString *const WMFSchemeHandlerArticleKeyQueryItem = @"articleKey";
+NSString *const WMFSchemeHandlerImageWidthQueryItem = @"imageWidth";
+
+static const NSInteger WMFCachedResponseCountLimit = 6;
+
+@interface WMFURLSchemeHandler ()
+@property (nonatomic, strong) WMFFIFOCache<NSString *, NSCachedURLResponse *> *responseCache;
+@property (nonatomic, strong) WMFFIFOCache<NSString *, MWKArticle *> *articleCache;
+@property (nonatomic, copy, nonnull) NSString *hostedFolderPath;
+@property (nonatomic, strong) NSURLSession *session;
+@property (nonatomic, strong) dispatch_queue_t activeTaskQueue;
+@property (nonatomic, strong) NSMutableSet *activeTasks;
+@end
+
+@implementation WMFURLSchemeHandler
+
++ (WMFURLSchemeHandler *)shared {
+    static dispatch_once_t onceToken;
+    static WMFURLSchemeHandler *shared;
+    dispatch_once(&onceToken, ^{
+        shared = [[WMFURLSchemeHandler alloc] init];
+    });
+    return shared;
+}
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        [self setup];
+    }
+    return self;
+}
+
+- (void)setup {
+    self.responseCache = [[WMFFIFOCache alloc] initWithCountLimit:WMFCachedResponseCountLimit];
+    self.articleCache = [[WMFFIFOCache alloc] initWithCountLimit:WMFCachedResponseCountLimit];
+    self.hostedFolderPath = [WikipediaAppUtils assetsPath];
+    self.session = [WMFSession urlSession];
+    self.activeTaskQueue = dispatch_queue_create("org.wikipedia.WKURLSchemeHandler.accessQueue", DISPATCH_QUEUE_CONCURRENT);
+    self.activeTasks = [[NSMutableSet alloc] init];
+}
+
+#pragma mark - Task handling
+
+- (BOOL)isTaskActive:(id <WKURLSchemeTask>)task {
+    __block BOOL isActive = NO;
+    if (!task) {
+        return isActive;
+    }
+    dispatch_sync(self.activeTaskQueue, ^{
+        isActive = [self.activeTasks containsObject:task];
+    });
+    return isActive;
+}
+
+- (void)finishTask:(id <WKURLSchemeTask>)task withResponse:(nullable NSHTTPURLResponse *)response data:(nullable NSData *)data error:(nullable NSError *)error {
+    if (![self isTaskActive:task]) {
+        return;
+    }
+    dispatch_barrier_async(self.activeTaskQueue, ^{
+        [self.activeTasks removeObject:task];
+    });
+    if (error) {
+        [task didFailWithError:error];
+    } else if (response) {
+        [task didReceiveResponse:response];
+        if (data) {
+            [task didReceiveData:data];
+        }
+        [task didFinish];
+    } else {
+        [task didFinish];
```

Comment:
I saved a couple articles then cold started the app with no internet connection and see a crash here when I try to load the saved article.